treegp/two_pcf.py

```diff
@@ -0,0 +1,410 @@
+                self.m.migrad()
+                results = [self.m.params[key].value for key in self.m.parameters]
+                self._fit_ok = self.m.accurate
+            else:
+                self.m = iminuit.Minuit.from_array_func(self.chi2, p0, print_level=0)
+                self.m.migrad()
+                results = [self.m.values[key] for key in self.m.values.keys()]
+                self._fit_ok = self.m.migrad_ok()
+
+        self._minuit_result = results
+        self.result = [np.sqrt(self.alpha[0][0]), results[0],
+                       results[1], results[2],
+                       self.alpha[1][0]]
+
+
+    def minimize_minuit(self, p0 = [3000., 0.2, 0.2]):
+        """
+        Launch the minimization by minuit given a set of starting point.
+        Launch again on different starting point if minuit failed to find
+        a solution.
+
+        :param p0: List of starting points.
+        """
+        self._minimize_minuit(p0=p0)
+
+        if not self._fit_ok:
+            N_restart = 3
+            g1 = np.linspace(-0.3, 0.3, N_restart)
+            size = np.linspace(p0[0] - p0[0]/10., 2*p0[0], N_restart)
+            g1, g2, size = np.meshgrid(g1, g1, size)
+            N_restart = int(N_restart**3)
+            g1 = g1.reshape(N_restart)
+            g2 = g2.reshape(N_restart)
+            size = size.reshape(N_restart)
+            for i in range(N_restart):
+                print("restart fit because failure")
+                new_p0 = [size[i], g1[i], g2[i]]
+                print(new_p0)
+                self._minimize_minuit(p0=new_p0)
+                if self._fit_ok:
+                    break
+        pcf = self._model_skl(self.result[0], self.result[1], 
+                              self.result[2], self.result[3])
+
+class two_pcf(object):
+    """
+    Fit statistical uncertaintie on two-point correlation function using bootstraping.
+
+    :param X:           Coordinates of the field.  (n_samples, 1 or 2)
+    :param y:           Values of the field. (n_samples)
+    :param y_err:       Error of y. (n_samples)
+    :param min_sep:     Minimum bin for treecorr. (float)
+    :param max_sep:     Maximum bin for treecorr. (float)
+    :param nbins:       Number of bins (if 1D correlation function) of the square root of the number
+                        of bins (if 2D correlation function) used in TreeCorr to compute the
+                        2-point correlation function. [default: 20]
+    :param anisotropic: 2D 2-point correlation function.
+                        Used 2D correlation function for the
+                        fiting part of the GP. (Boolean)
+    :param robust_fit:  Used minuit to fit hyperparameter. Works only
+                        anisotropic is True. (Boolean)
+    """
+    def __init__(self, X, y, y_err,
+                 min_sep, max_sep, nbins=20, 
+                 anisotropic=False, robust_fit=False, 
+                 p0=[3000., 0., 0.]):
```

Comment:
In other places `p0` defaults to `[3000., 0.2, 0.2]`---is there a reason for both cases in this code, or should either be changed so that everything defaults to some consistent set of values?

Reply:
To be honest, I don't remember why I put in some part of the code the value `[3000., 0.2, 0.2]`. I guess to start with `[1, 0, 0]` would be better.